### PR TITLE
[Formerly #481] False positive for long running response status code

### PR DESCRIPTION
**Pull request by bdefoy**
Refactor `LongRunningResponeStatusCode` rule and docs for clarity and to check for the appropriate ARM LRO response codes.

Formerly #481

### CODE REVIEW

**Comment by konrad-jamrozik** [comment]
I would appreciate a proper PR description summarizing the changes made. The description should be in this PR, not the previous one. Plus, the previous PR doesn't have any description anyway.

File: docs/long-running-response-status-code.md [hunk 1]
```diff
@@ -18,13 +18,28 @@ A '{0}' operation '{1}' with x-ms-long-running-operation extension must have a v
...
 ## Description
 
-For ARM spec, the allowed response status codes for a long DELETE operation are "200" & "204"; the allowed response status codes for a POST operation are "200", "201" ,"202", & "204"; the allowed response status codes for a PUT/PATCH operation are "200" & "201".
-For Data plane spec, the allowed response status codes for a long DELETE operation are "200","202", & "204"; the allowed response status codes for a POST operation are "200", "201" ,"202", & "204"; the allowed response status codes for a PUT/PATCH operation are "200","201", & "202".
+For **ARM (Control Plane) specs**, valid response status codes for long-running (LRO) operations are as follows:
+
+| Operation Name | Valid Response Codes           |
+| -------------- | ------------------------------ |
+| Delete         | `200` and `204`                |
+| Post           | `200`, `201`, `202`, and `204` |
```
Comment (rkmanda) [inline docs/long-running-response-status-code.md:26]: >`201 [](http://example.com/codeflow?start=26&length=4)

201 is not a valid status code for Post

File: docs/rules.md [hunk 2]
```diff
@@ -451,23 +451,23 @@ Please refer to [long-running-response-status-code-data-plane.md](./long-running
 
 ### LongRunningResponseStatusCode
 
-For ARM spec, the allowed response status codes for a long DELETE operation are "200" & "204"; the allowed response status codes for a POST operation are "200", "201" ,"202", & "204"; the allowed response status codes for a PUT/PATCH operation are "200" & "201".
-For Data plane spec, the allowed response status codes for a long DELETE operation are "200","202", & "204"; the allowed response status codes for a POST operation are "200", "201" ,"202", & "204"; the allowed response status codes for a PUT/PATCH operation are "200","201", & "202".
+For **ARM (Control Plane) specs**, valid response status codes for long-running (LRO) operations are as follows:
+| Operation Name | Valid Response Codes           |
+| -------------- | ------------------------------ |
+| Delete         | `200` and `204`                |
+| Post           | `200`, `201`, `202`, and `204` |
```
Comment (rkmanda) [inline docs/rules.md:458]: >`201` [](http://example.com/codeflow?start=26&length=5)

remove 201

File: packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts [hunk 3]
```diff
@@ -1,50 +1,76 @@
...
-const longRunningResponseStatusCode = (methodOp: any, _opts: any, ctx: any, validResponseCodesList: any) => {
-  if (methodOp === null || typeof methodOp !== "object") {
+// operations that this rule evaluates
+const RULE_OPERATION_NAMES = ["delete", "put", "patch", "post"]
+// extension to indicate a long-running operation (LRO)
+const LRO_EXTENSION = "x-ms-long-running-operation"
+// valid response codes for long-running (LRO) ARM (Control Plane) operations
+const ARM_LRO_RESPONSE_CODES_BY_OPERATION: any = {
+  delete: ["200", "204"],
+  post: ["200", "201", "202", "204"],
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts:10]: >201 [](http://example.com/codeflow?start=17&length=3)

remove

File: packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts [hunk 4]
```diff
@@ -1,50 +1,76 @@
...
 
-const longRunningResponseStatusCode = (methodOp: any, _opts: any, ctx: any, validResponseCodesList: any) => {
-  if (methodOp === null || typeof methodOp !== "object") {
+// operations that this rule evaluates
+const RULE_OPERATION_NAMES = ["delete", "put", "patch", "post"]
+// extension to indicate a long-running operation (LRO)
+const LRO_EXTENSION = "x-ms-long-running-operation"
+// valid response codes for long-running (LRO) ARM (Control Plane) operations
+const ARM_LRO_RESPONSE_CODES_BY_OPERATION: any = {
+  delete: ["200", "204"],
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts:9]: >] [](http://example.com/codeflow?start=23&length=1)

and 202

File: packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts [hunk 5]
```diff
@@ -1,50 +1,76 @@
...
+  post: ["200", "201", "202", "204"],
+  put: ["200", "201"],
+  patch: ["200", "201", "202"],
```
Comment (rkmanda) [inline packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts:12]: >201 [](http://example.com/codeflow?start=18&length=3)

no 201 for patch

**Comment by rkmanda** [comment]
Whats the key issue being solved with this PR? Can u describe what the false negative is?

---
In reply to: [1502611483](https://github.com/Azure/azure-openapi-validator/pull/491#issuecomment-1502611483) [](http://example.com/codeflow?ancestors=1502611483,1502611483)

**Review by rkmanda — CHANGES_REQUESTED** [review]
:clock1:

File: docs/long-running-response-status-code.md (hunk 1 above)
Comment (TimLovellSmith) [reply docs/long-running-response-status-code.md:26]: I thought it is:

https://stackoverflow.com/questions/1860645/create-request-with-post-which-response-codes-200-or-201-and-content

File: packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts (hunk 5 above)
Comment (TimLovellSmith) [reply packages/rulesets/src/spectral/functions/Extensions/long-running-response-status-code.ts:12]: I seem to recall I've queried people about whether PATCH creating resources makes sense before and the consensus seemed to be no... it should be used only for updates. I already forgot why though, sigh.

File: docs/long-running-response-status-code.md (hunk 1 above)
Comment (bdefoy) [reply docs/long-running-response-status-code.md:26]: Consolidating all the review comments here. For context, I only converted the existing documentation into a table and went with that for the rule. However, for control plane do we want it to be this instead?

| Operation Name | Valid Response Codes           |
| -------------- | ------------------------------ |
| LRO Delete         | `200`, `202`, and `204`                |
| LRO Post           | `200`, `202`, and `204` |
| LRO Put            | `200` and `201`                |
| LRO Patch          | `200`               |

Changes:
- Added `202` for LRO delete
- Removed `201` for LRO post
- Removed `201` for LRO patch

**Review by konrad-jamrozik — APPROVED** [review]

**Review by konrad-jamrozik — CHANGES_REQUESTED** [review]
Requesting the stuff I wrote here:
https://github.com/Azure/azure-openapi-validator/pull/491#issuecomment-1502611483

**Comment by TimLovellSmith** [comment]
Any changes needed to 'DeleteOperationAsyncResponseValidation' ?

Today it says 
'An async DELETE operation response include status code 202 with 'Location' header. Must support status code 200 if operation can be completed synchronously. Must support 204 (resource doesn't exists). Operation must also add "x-ms-long-running-operation and x-ms-long-running-operation-options" to mark that it is a long running operation (in case of 202) and how it is tracked (Location header).'

This, I agree with: >x-ms-long-running-operation

But saying this *must* be added in the same sentence, doesn't sound right: > x-ms-long-running-operation-options

**Pull request closed without merging**